### ISSUE TITLE
Conditional Single/Multiline enforcing based on length

### ISSUE THREAD

**Issue by MadLittleMods** [stylelint/stylelint#409]
Currently in my code the switch from single to multiline happens based on my subjective opinion of the length. I would love some linting rules to enforce single vs multi line based on the value length.

This concept applies would apply to:

 - Comments
      - `/* */`
 - Function parameters
      - `linear-gradient(45deg, rgba(215, 215, 215, 1) 0%);`
 - Value list
      - `transition: color 0.1s ease, background-color 0.3s linear, opacity 0.5s ease;`
 - Media Queries
      - `@media screen and (color), projection and (color) {}`

**Comment by jeddy3** [comment]
Thanks, as always, for the suggestions :)

Way back, there was talk of a `rule-single-line-max-length` rule. It was a low priority, but it's still kicking around in #1 and waiting to be tackled.

Your suggestion sounds like something similar, but applied to comments, function lists, value lists, and media query lists respectively:

* `comment-single-line-max-length: int`
* `function-list-single-line-max-length: int`
* `value-list-single-line-max-length: int`
* `media-query-list-single-line-max-length: int`

These feel like worthwhile rules to me. Although, we should probably discuss if these rules are niche enough to be consider good candidates for plugins (and wait for @davidtheclark to weigh into that discussion). However, if you'd like a make a start on this, such a discussion shouldn't impact you much as you can migrate any work you do out into a plugin, if that's what we all decide is best.

Do you think those four rules will allow you to enforce your desired code style?

**Comment by MadLittleMods** [comment]
I think those rules should be sufficient.

A couple notes:

 - For a value-list, I think we should limit it to each value as I wouldn't want some long value to be cut into multiple lines.
      - Same applies for function parameters

**Comment by jeddy3** [comment]
Yep, that makes sense. Although, I thought such a possibility is accounted for already because we're making the rule specific to the max length of _single lines_. Once the list spans multiple lines, it is no longer check by the rule.

For example, say we set `value-list-single-line-max-length: 60`

The following CSS will produce a warning:

```css
a {
  background-image: url("reaaalllllllllllllllllyyyyyyy-llloonngggggg-value-of-72-chars.png"), url(second-value.png);
}
```

It is _a list_, it is on a _single line_ and the list exceeds 60 characters. 

So the author changes the CSS to:

```css
a {
  background-image: 
    url("reaaalllllllllllllllllyyyyyyy-llloonngggggg-value-of-72-chars.png"), 
    url(second-value.png);
}
```

The rule doesn't produce a warning this time. It is still _a list_ and it still exceeds 60 characters, but it now *span multiple lines*.

Likewise, the following won't produce a warning:

```css
a {
  background-image: url("reaaalllllllllllllllllyyyyyyy-llloonngggggg-value-of-72-chars.png");
}
```

It exceeds 60 characters and is _single line_, but is is _not_ a list.

Is that what you meant? Or have I gotten the wrong end of the stick? :)

**Comment by MadLittleMods** [comment]
@jeddy3 Correct.

---

For comments, it will be strict length max limit no matter if single or multi-line where you always have to wrap before the max character limit.

---

Media queries should also be broken up even if spanning multiple lines but are safe if a single expression and combinator extend beyond that limit.

Perhaps the combinator being able to be on the same line should be an option. I don't think this should be split out to another rule as this one needs to pass somehow.

```
media-query-list-single-line-max-length: [2, 50, {
    combinator: 'always-same-line' || 'never'
}];
```

Example where the limit was 50 characters:
```css
@media screen and (max-width: 1234567890001234567890px),
	projection and
	(max-width: 12345678900012345678908888555554444422233344411145566677788px) and
	(max-width: 1234567890001234567890px) and
	(max-width: 1234567890001234567890px) and
	(max-width: 1234567890001234567890px) {

}
```

**Comment by jeddy3** [comment]
Eeek, this could be a can a worms. 

I think I'm now leaning towards having a single `max-line-length` rule in core that applies to the sheet, and pushing these more specific use-case into plugins.

You can then experiment with options and rule names in those plugins, until you find something that fits for you.

Just throwing some suggestions out there for those plugins though...

I'm not sure the name `media-query-list-single-line-max-length` makes sense in the context of these new options as we're now factoring in the "and" of media queries (where as `media-query-list` is specific to media queries separated by commas ([ref](http://www.w3.org/TR/mediaqueries-4/#mq-syntax)):

Perhaps if you bring the _thing_ up to the statement level i.e. `at-media-rule`, and then you can create the necessary options to control the length of the _at media statement_. e.g. 

```js
`at-media-rule-max-line-length: [2, 50, {
    breakOnMediaQueryListComma: true|false,
    breakOnMediaQueryAnd: true|false
}];
```

For consistency, you can could also have `rule-max-line-length: int` instead of `value-list-single-line-max-length` and `function-list-single-line-max-length`, as that's also at the statement level.

And `comment-max-line-length` instead of `comment-single-line-max-length`.

Does that make sense?

But, like I mentioned above, these might still be too specific a use case to be included in the core. And so, I think make perfect candidates for plugins.

**Comment by MadLittleMods** [comment]
`max-line-length` is nice but I feel the many will run into its shortcomings.

A plugin is great but requires extra "research" to find the solution.

**Comment by jeddy3** [comment]
> `max-line-length` is nice but I feel the many will run into its shortcomings. A plugin is great but requires extra "research" to find the solution.

Fair points :) 

We're looking to find a balance (similar to how [eslint must do the same](http://eslint.org/docs/developer-guide/contributing/new-rules#core-rule-guidelines)) between the extra time spent finding plugin rules and the time needed by new users to familiarising themselves with the rules in core. 

I guess one approach to making a more informed decision on where that balance might lie, would be to start these rules as plugins and if they are popular and/or we get more requests for that functionality, we can roll them into core.

Anyway, before settling on an approach for now. Lets wait for @davidtheclark to weigh in, as he might have some good reasons we've not thought of for including the rules in core from the outset :)

**Comment by davidtheclark** [comment]
What would be the benefits of splitting this up into four separate rules? I am having a hard time imagining why somebody would not just set a single line-length limit, as I think other linters do.

**Comment by MadLittleMods** [comment]
@davidtheclark One rule would work for me.

But I would still want it to account for not breaking on a value-list item, function parameter or media query expression+combinator. This could be behind a `{ strict: false }` option.

**Comment by davidtheclark** [comment]
@MadLittleMods : Can you give some specific examples of the contexts in which `max-line-length` with a single value would not be sufficient? I've gotten lost in the discussion.

**Comment by MadLittleMods** [comment]
@davidtheclark I think these two messages sum it up well. Bottom-line is that it needs to cover the case where a line-break would actually break the CSS.

 - https://github.com/stylelint/stylelint/issues/409#issuecomment-146647730
 - https://github.com/stylelint/stylelint/issues/409#issuecomment-146651523

**Comment by jeddy3** [comment]
I think we should close this issue in favour of #423 and using the special disable comments if a line is too long, for example if you have a long url e.g.

```css
a {
  background-image: 
    url("reaaalllllllllllllllllyyyyyyy-llloonngggggg-value-of-72-chars.png"), 
    url(second-value.png);
}
```

Plugins are be used if for anything more bespoke. Thoughts @davidtheclark ?

**Comment by davidtheclark** [comment]
Sure. 

I think it's also feasible that we could throw in a qualification so that if the line actually has no spaces (except at the beginning, around a colon, comma, or `and`) it is ignored. That might address the problems @MadLittleMods anticipates without demanding disable comments. Can we think of any reason that *wouldn't* work -- any other problems that exception would raise?

**Comment by MadLittleMods** [comment]
@davidtheclark This sounds like a solid solution for the time-being.

**Comment by jeddy3** [comment]
That's a nice solution :)

> Can we think of any reason that wouldn't work -- any other problems that exception would raise?

Not off hand. It feels like one of those things that'll only pop up when we come to implement it.

**Comment by ramiAbdou** [comment]
What is the status on this? I would really love for my SCSS to be fixed to one line if it is less than my max line length...
